src/man/clockwork.1
.TH "clockwork" "1" "0.99.3" "Nigel Stuckey, System Garden Ltd" "Habitat"
.SH "NAME"
.LP 

clockwork \- collection daemon for the Habitat suite
.SH "SYNTAX"
.LP 
clockwork [\fI\-c <purl>\fP] [\fI\-C <cfcmd>\fP] 
[\fI\-e <fmt>\fP] [\fI\-dDhsv\fP] [\fI\-j <jobs>\fP]
.SH "DESCRIPTION"
.LP 
Clockwork is the local agent for the Habitat suite.
It runs as a daemon process on each machine to be monitored and is designed
to carry out data collection, log file monitoring, data\-driven actions and 
the distribution of collected data.
.LP 
The default jobs are to collect system, network, storage and uptime
statistics on the local machine and make them available in a standard place.
The collection of process data and file monitoring is available by
configuring the jobs that drive clockwork.
Configuration can be carried out at a local, regional and global level
to allow deligation.
One public and many private instances of clockwork 
can exist on a single 
machine, allowing individual users to carry out custom data collection 
Data is normally held in ring buffers or queues on the local machine
using custom datastores to be self contained and scalable.
Periodic replication of data rings to a repository is used for archiving
and may be done in reverse for central data transmission.
.SH "OPTIONS"
.LP 
.TP 
\fB\-c\fR \fI<purl>\fP
Append user configuration data from the route <purl>, rather than the default 
file ~/.habrc.
.TP 
\fB\-C\fR \fI<cfcmd>\fP
Append a list of configuration directives from <cfcmd>, separated by semicolons.
.TP 
\fB\-d\fR
Place \fBclockwork\fR in diagnostic mode, giving an additional 
level of logging and sending the text to stderr rather than the default 
or configured destinations.
In daemon mode, will send output to the controlling terminal.
.TP 
\fB\-D\fR
Place \fBclockwork\fR in debug mode. 
As \-d above but generating a great deal more information, designed to be 
used in conjunction with the source code. 
Also overrides normal outputs and will send the text to stderr.
In daemon mode, will send output to the controlling terminal.
.TP 
\fB\-e\fR \fI<fmt>\fP
Change the logging output to one of eight preset alternative formats, 
some showing additional information.
<fmt> must be 0\-7.
See LOGGING below.
.TP 
\fB\-h\fR
Print a help message to stdout and exit
.TP 
\fB\-v\fR
Print the version to stdout and exit
.TP 
\fB\-j\fR \fI<jobs>\fP
Override public job table with a private one provided by the route
<jobs>.
\fBClockwork\fR will not daemonise, run a data service or take an 
exclusive system lock (there can only be one public \fBclockwork\fR
instance).
Implies \-s and alters the logging output to stderr, unless overridden
with the range of \fBelog\fR configuration directives.
.TP 
\fB\-s\fR
Disable the public data service from being run, but will continue to save data 
as dictated by configuration.
.SH "DEFAULTS"
.LP 
When clockwork starts it reads $HAB/etc/habitat.conf and ~/.habrc for
configuration data (see CONFIGURATION for more details).
Unless overridden, clockwork will then look for its jobs inside the default
public datastore for that machine, held in \fI$HAB/var/<hostname>.rs\fR
(the route address is \fIrs:$HAB/var/<hostname>.rs,jobs,0\fR, see below for 
an explanation).
If it does not find the jobs, clockwork bootstaps itself by copying a
default job template from the file \fI$HAB/lib/clockwork.jobs\fR into the 
public datastore and then carries on using the datastore version.
.LP 
The default jobs run system, network and storage data gathering probes
every 60 seconds.
It saves results to the public datastore using the template route 
\fIrs:$HAB/var/<hostname>.rs,<jobname>,60\fR and errors to 
\fIrs:$HAB/var/<hostname>.rs,err_<jobname>,60\fR
.LP 
All other errors are placed in \fIrs:$HAB/var/<hostname>.rs,log,0\fR
.SH "ROUTES"
.LP 
To move data around in clockwork, an enhanced URL is used as a form of 
addressing and is called a 'route' (also known as a pseudo\-url or p\-url 
in documentation).
The format is <driver>:<address>, where driver must be 
one of the following:\-
.TP 
\fBfile:\fR \fBfileov:\fR
reads and write to paths on the filesystem.
The format is file:<file path>, which will always append text to the
file when writing.
The fileov: driver will overwrite text when first writing and is 
suitable for configuration files or states.
.TP 
\fBhttp:\fR \fBhttps:\fR
reads and writes using HTTP or HTTPS to a network address.
The address is the server name and object name as a normal URL convention.
.TP 
\fBrs:\fR
read and writes to a ring store, the primary local storage mechanism.
Tabular data is stored in a time series in a queue or ring buffer 
structure.
Multiple rings of data can be stored in a single ringstore file, using
different names and durations.
.TP 
\fBsqlrs:\fR
reads and writes tabular data to a remote repository service using the 
SQL Ringstore method, which is implemented over the HTTP protocol.
Harvest provides repository services.
Stores tabular data in a time series, addressed by host name, ring name
and duration.
Data is stored in a wueue or ring buffer storage.
.SH "CONFIGURATION"
By default, \fBclockwork\fR will collect system, network and storage statistics for
the system on which it runs. 
All the data is read and written from a local datastore, apart from 
configuration items which come from external sources.
These external configuration sources govern the operation of all the habitat
commands and applications.

Refer to the \fIhabconf\fR(5) man page for more details.
.SH "JOB DEFINITIONS"
.LP 
Jobs are defined in a multi columned text format, headed by the magic 
string 'job 1'.
Comments may appear anywhere, starting with '#' and running to the end 
of the line.
.LP 
Each job is defined on a single line containing 11 arguments, which in
order are:\-
.TP 
1. start
when to start the job, in seconds from the starting of clockwork
.TP 
2. period
how often to repeat the job, in seconds
.TP 
3. phase
not yet implemented
.TP 
4. count
how many times the job should be run, with 0 repeating forever
.TP 
5. name
name of the job
.TP 
6. requester
who requested the job, by convention the email address
.TP 
7. results
the route where results should be sent
.TP 
8. errors
the route where errors should be sent
.TP 
9. nslots
the number of slots created in the 'results' and 'errors' routes, 
if applicable (applies to timestore and tablestore).
.TP 
10.method
the job method
.TP 
11.command
the arguments given to each method
.LP 
See the \fBhabmeth(1)\fR manpage for details of the possible methods that may be 
specified and the commands that can accept.
.SH "DATA ORGANISATION"
.LP 
Data is stored in sequences of tabular information.
All data has an ordered independently of time, allowing multiple separate
samples that share the same time interval.
This data is stored in a ringbuffer, which allows data to grow to a certain 
number of samples before the oldest are removed and their space recycled.
Throughout the documentation, each collection of samples is known as
a \fBring\fR, and may be configured to be a simple queue, where data 
management is left up to administrators.
.LP 
To limit the amount of storage used, data in a ring can be sampled 
periodically to form new summary data and stored in a new ring with 
a different period.
In \fBhabitat\fR, this is known as \fBcascading\fR and takes place on 
all the default collection rings.
Several levels of cascading can take place over several new rings, 
This allows summaries at different frequencies to be collected and tuned
to local requirements.
.LP 
See the \fBhabmeth\fR(1) man page for more information about the \fBcascade\fR method.
.SH "DATA REPLICATION"
.LP 
Any ring of information can be sent to or from the repository at 
known intervals, allowing a deterministic way of updating both repository 
and collection agent.
.LP 
This is implemented as a regular job which runs the \fBreplicate\fR method.
Data for the method is provided by configuration parameters which can be 
set and altered in the organisation. 
Thus the replication job does not normally need to be altered to change 
the behaviour.
.LP 
See the \fBhabmeth\fR(1) man page for the replicate method and the 
formation of the configuration data.
.SH "LOGGING"
.LP 
\fBClockwork\fR and the probes that provide data, also generate 
information and error messages. By convention, these are stored in the 
route specification \fIts:$hab/var/<host>.ts,log\fR
The convention for probes is to store their errors in
\fIts:$HAB/var/\fB<host>\fR.ts,e.\fB<jobname>\fR\fR.

To override the logging location, use the range of \fBelog\fR 
configuration directives, or rely on the options \-d, \-D, \-j, 
which will alter the location to stderr as a side effect.
See habconf(5) for details.
Probe logging is configurable for each job in the job table.

The logging format can be customised using one of a set of configuration
directives (see habconf(5)).
For convenience, the \-e flag specifies one of eight preconfigured
text formats that will be sent to the configured location:\-
.TP 
0 
all 17 possible log variables
.TP 
1 
severity character & text
.TP 
2 
severity & text
.TP 
3 
severity, text, file, function & line
.TP 
4 
long severity, short time, short program name, file, function, line & text
.TP 
5 
date time, severity, long program name, process id, file, function, line, origin, code & text
.TP 
6 
unix ctime, seconds since 1970, short program name, process id, thread id, file, function, line, origin, code & text
.TP 
7 
severity, file, line, origin, code, text
.SH "FILES"
.LP 
If run from a single directory $HAB:\-
.br 
\fI$HAB/bin/clockwork\fP
.br 
\fI$HAB/var/<hostname>.rs\fP, \fI$HAB/lib/clockwork.jobs\fP
.br 
\fI/tmp/clockwork.run\fP
.br 
\fI~/.habrc\fP, \fI$HAB/etc/habitat.conf\fP
.LP 
If run from installed Linux locations:\-
.br 
\fI/usr/bin/habitat\fP
.br 
\fI/var/lib/habitat/<hostname>.rs\fP, \fI/usr/lib/habitat/clockwork.jobs\fP
.br 
\fI/var/lib/habitat/clockwork.run\fP
.br 
\fI~/.habrc\fP, \fI/etc/habitat.conf\fP
.SH "ENVIRONMENT VARIABLES"
.LP 

.SH "EXAMPLES"
.LP 
Type the following to run \fBclockwork\fR in the standard way.
This assumes it is providing public data using the standard job file, 
storing in a known place and using the standard network port for the
data service.

clockwork

On a more secure system, you can prevent the data service from being started

clockwork \-s

Alternatively you can run it in a private mode by specifying '\-j' and a
replacement job file.

clockwork \-j <file>
.SH "AUTHORS"
.LP 
Nigel Stuckey <nigel.stuckey@systemgarden.com>
.SH "SEE ALSO"
.LP 
killclock(1), ghabitat(1), habget(1), habput(1), irs(1), 
habedit(1), habprobe(1), habmeth(1), habconf(5)


src/cmd/clockwork.c
/*
 * Clockwork - a periodic execution utility for iiab
 *
 * Nigel Stuckey, December 1998
 * Copyright System Garden Limited 1998-2004. All rights reserved.
 *
 * Think of clockwork as cron with `knobs on'.
 * Using functionality in libiiab, it has more timing control
 * than cron and has the ability to execute with arbitary execution types.
 * It also pipes its jobs through the route_ class, thus using timestore
 * from stdout of a job.
 */

#include <stdio.h>
#include <unistd.h>
#include <time.h>
#include <stdlib.h>
#include "../iiab/route.h"
#include "../iiab/itree.h"
#include "../iiab/tree.h"
#include "../iiab/cf.h"
#include "../iiab/util.h"
#include "../iiab/meth.h"
#include "../iiab/elog.h"
#include "../iiab/iiab.h"
#include "../iiab/job.h"
#include "../iiab/nmalloc.h"
#include "../iiab/httpd.h"
#include "../iiab/sig.h"
#include "../iiab/runq.h"
#include "../iiab/job.h"
#include "../probe/probe.h"

int main(int, char **);
int load_jobs(char *purl);
void stopclock(int sig /* signal vector */);

/* initialise globals */
char usagetxt[] = "[-j <jobs>] [-s]\n" \
     "where -j <jobs>   use jobs from route <jobs>; don't daemonise (imply -s)\n"
     "      -s          server off: do not listen for data requests from network";
char helptxt[] = "no help for the weary";
char *cfdefaults = 
     "iiab.debug -1\n"
     "job.debug  -1\n"
     "nmalloc    0\n"	/* 0: memory checking off, !0: memory checking on */
     "log        stderr:\n"
     "jobs       rs:%s/%%h.rs,clockwork,0\n"	/*file:clockwork.jobs*/
     "elog.all   none:\n"
     "elog.above warning rs:%s/%%h.rs,log\n";
char *cfdefaults2 = 
     "iiab.debug -1\n"
     "job.debug  -1\n"
     "nmalloc    0\n"	/* 0: memory checking off, !0: memory checking on */
     "log        stderr:\n"
     "elog.all   none:\n"
     "elog.above warning stderr:\n";
int   debug=0;			/* Debug state */
int   clock_done_init=0;

#define CLOCKWORK_KEYNAME "clockwork"
#define DEFJOBS_TYPE "file:"
#define DEFJOBS_FILE "clockwork.jobs"

int main(int argc, char **argv) {
     int r, joblen, errorstatus=0;
     char *jobpurl, jobpurl_t[1024], *jobtxt, *defjobs, buf[500];
     time_t clock;
     ROUTE jobrt;

     /* initialise, fetching the directory locations before hand and
      * then ensure we are the only clockwork running on this box.
      * By default send errors to the default data store, unless 
      * errors are overridden or -j switch is given. */
     iiab_dir_locations(argv[0]);
     if (iiab_iscmdopt("j", argc, argv))
          snprintf(buf, 500, cfdefaults2);
     else
          snprintf(buf, 500, cfdefaults, iiab_dir_var, iiab_dir_var);
     iiab_start("sj:", argc, argv, usagetxt, buf);

     /* process switches and arguments */
     /* debug flag */
     if (cf_defined(iiab_cf, "d") && cf_getint(iiab_cf, "d") == -1)
	  debug++;

     /* initialise the classes needed in addition to those started by 
      * iiab_start(); mostly these are for job & dispatching.
      */
     sig_init();
     meth_init();
     meth_add(&probe_cbinfo);
     runq_init(time(NULL));
     job_init();
     clock_done_init++;
     if (cf_defined(iiab_cf, "j")) {
          /* replacement jobs and private mode */
          cf_putstr(iiab_cf, "jobs", cf_getstr(iiab_cf, "j"));
     } else {
	  /* default running: we want to be a daemon! */
	  iiab_daemonise();
	  iiab_lockordie(CLOCKWORK_KEYNAME);

	  /* only in daemon mode can we provide the server unless its 
	   * turned off */
	  if ( ! cf_defined(iiab_cf, "s") ) {
	       /* start http server with the following services */
	       httpd_init();
	       httpd_addpath("/ping",     httpd_builtin_ping);
	       httpd_addpath("/cf",       httpd_builtin_cf);
	       httpd_addpath("/elog",     httpd_builtin_elog);
	       httpd_addpath("/info",     httpd_builtin_info);
	       httpd_addpath("/local/",   httpd_builtin_local);
	       httpd_addpath("/localtsv/",httpd_builtin_local);
	       httpd_start();
	  }
     }

     /* set up signal handlers */
     sig_setexit(stopclock);

     /* check 'jobs' variable exists, if not we can't set one up */
     jobpurl = cf_getstr(iiab_cf, "jobs");
     route_expand(jobpurl_t, jobpurl, "NOJOB", 0);
     if (jobpurl_t == NULL || jobpurl_t == '\0') {
	  elog_printf(FATAL, "unable to load jobs, as the "
		      "configuration directive was not set. Please set "
		      "the directive `jobs' in the configuration file "
		      "to the route containing jobs. For example, "
		      "`jobs=file:/etc/clockwork.jobs' will look for the "
		      "file /etc/clockwork.jobs. If the route is set but "
		      "does not exist, then it will be created using the "
		      "default job definitions");
	  errorstatus = 1;
	  goto end_app;
     }

     /* Access the job table to see if it exists. If it does not, then
      * assume that thats were we want the default job config created */
     if ( route_access(jobpurl_t, NULL, ROUTE_READOK) != 1 ) {
	  /* jobs directive set but no file there => create it */

	  /* read the default jobs file and save them in the standard
	   * place. This is normally the data store, but could be 
	   * a file */
	  defjobs = util_strjoin(DEFJOBS_TYPE, iiab_dir_lib, "/",
				 DEFJOBS_FILE, NULL);
	  jobtxt = route_read(defjobs, NULL, &joblen);
	  if (jobtxt == NULL) {
	       elog_printf(FATAL, "unable to read default job table. "
			   "Either (1) create job table in expected place "
			   "(%s) so avoiding use of defaults or (2) make "
			   "default jobs readable/available (%s)",
			   jobpurl_t, defjobs);
	       errorstatus = 2;
	       goto end_app;
	  }

	  jobrt = route_open(jobpurl_t, "Job table for clockwork", NULL, 10);
	  if (jobrt == NULL) {
	       elog_printf(FATAL, "unable to create initial job table "
			   "to write defaults. Please make job table "
			   "writable (%s) or create manually",
			   jobpurl_t);
	       nfree(jobtxt);
	       errorstatus = 3;
	       goto end_app;
	  }

	  r = route_write(jobrt, jobtxt, joblen);
	  if (r == -1) {
	       elog_printf(FATAL, "unable to create initial job table "
			   "to write defaults. Please make job table"
			   "writable (%s) or create manually",
			   jobpurl_t);
	       route_close(jobrt);
	       nfree(jobtxt);
	       errorstatus = 4;
	       goto end_app;
	  }

	  elog_printf(INFO, "no jobs found in %s: created default", 
		      jobpurl_t);
	  route_close(jobrt);
	  nfree(jobtxt);
     }

     /* load jobs */
     r = job_loadroute( jobpurl_t );
     if (r == -1)
          elog_die(FATAL, "failed to load jobs from %s",
		   cf_getstr(iiab_cf, "jobs") );
     else
          elog_printf(INFO, "loaded %d jobs", r);

     /* run jobs in var if we have a public responsibility */
     if ( ! cf_defined(iiab_cf, "j") )
	  chdir(iiab_dir_var);
     while(1) {
          elog_printf(DEBUG, "relay returns %d", meth_relay());
	  /*runq_dump();*/
     }

     /* shut down and clear up */
 end_app:     
     job_fini();
     runq_fini();
     meth_fini();
     iiab_stop();
     if (errorstatus) {
	  clock = time(NULL);
	  fprintf(stderr, "%s: exit with errorstatus %d at %s", argv[0], 
		  errorstatus, ctime(&clock));
     }
     exit(errorstatus);
}


/* shutdown clockwork */
void stopclock(int sig /* signal vector */) {
     int r;
     time_t clock;

     if (!clock_done_init)
          _exit(r);

     sig_off();
     elog_printf(INFO, "clockwork shutting down from signal %d", sig);

     runq_disable();
     r = meth_shutdown();
     if (r) {
	  clock = time(NULL);
          elog_printf(WARNING, "%d jobs did not shutdown normally", r);
	  fprintf(stderr, "%s: shutdown from signal %d, meth_shutdown() %d "
		  "at %s", cf_getstr(iiab_cmdarg, "argv0"), sig, 
		  r, ctime(&clock));
     } else
          elog_printf(INFO, "%s successfully shutdown", 
		      cf_getstr(iiab_cmdarg, "argv0"));
     /*job_clear();*/

     /* shut down and clear up */
     job_fini();
     runq_fini();
     meth_fini();
     iiab_stop();
     exit(r);
}

